presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PropertyDerivations.java

```diff
@@ -433,20 +433,26 @@ public ActualProperties visitJoin(JoinNode node, List<ActualProperties> inputPro
                             .unordered(true)
                             .build();
                 case FULL:
-                    if (probeProperties.getNodePartitioning().isPresent()) {
-                        Partitioning nodePartitioning = probeProperties.getNodePartitioning().get();
-                        ImmutableList.Builder<Expression> coalesceExpressions = ImmutableList.builder();
-                        for (Symbol column : nodePartitioning.getColumns()) {
-                            for (JoinNode.EquiJoinClause equality : node.getCriteria()) {
-                                if (equality.getLeft().equals(column) || equality.getRight().equals(column)) {
-                                    coalesceExpressions.add(new CoalesceExpression(ImmutableList.of(equality.getLeft().toSymbolReference(), equality.getRight().toSymbolReference())));
+                    if (probeProperties.getNodePartitioning().isPresent() && buildProperties.getNodePartitioning().isPresent()) {
+                        Partitioning probePartitioning = probeProperties.getNodePartitioning().get();
+                        Partitioning buildPartitioning = buildProperties.getNodePartitioning().get();
+                        if (probePartitioning.isOnlyPartitionedOnColumns() && buildPartitioning.isOnlyPartitionedOnColumns()) {
+                            ImmutableList.Builder<Expression> coalesceExpressionsBuilder = ImmutableList.builder();
+                            for (Symbol column : probePartitioning.getColumns()) {
+                                for (JoinNode.EquiJoinClause equality : node.getCriteria()) {
+                                    if (equality.getLeft().equals(column) || equality.getRight().equals(column)) {
+                                        coalesceExpressionsBuilder.add(new CoalesceExpression(ImmutableList.of(equality.getLeft().toSymbolReference(), equality.getRight().toSymbolReference())));
+                                    }
                                 }
                             }
+                            List<Expression> coalesceExpressions = coalesceExpressionsBuilder.build();
+                            // If the join clause includes equi-join of all partition symbols, the result is partitioned on coalesce of all partition symbols
+                            if (probePartitioning.getColumns().size() == coalesceExpressions.size()) {
```

Comment:
Potentially redundant because of how you construct this, but an explicitly checking build side would also be good.

I am not sure, but I believe broadcast could trigger a mismatch.